services/src/main/scala/cromwell/pipeline/service/ProjectService.scala

```diff
@@ -12,22 +13,59 @@ class ProjectService(projectRepository: ProjectRepository)(implicit executionCon
   def getProjectById(projectId: ProjectId): Future[Option[Project]] =
     projectRepository.getProjectById(projectId)
 
-  def addProject(request: ProjectAdditionRequest): Future[ProjectId] = {
+  def getProjectByName(namePattern: String, userId: UserId): Future[Option[Project]] = {
+    val result = projectRepository.getProjectByName(namePattern)
+    result.flatMap {
+      case Some(project) if project.ownerId == userId => result
+      case Some(_)                                    => Future.failed(new ProjectAccessDeniedException)
+      case None                                       => Future.failed(new ProjectNotFoundException)
+    }
+  }
+
+  def addProject(request: ProjectAdditionRequest, userId: UserId): Future[ProjectId] = {
     val project =
       Project(
         projectId = ProjectId(UUID.randomUUID().toString),
-        ownerId = request.ownerId,
+        ownerId = userId,
         name = request.name,
-        repository = request.repository,
+        repository = "test_repo",
         active = true
       )
     projectRepository.addProject(project)
   }
 
-  def deactivateProjectById(projectId: ProjectId): Future[Option[Project]] =
-    for {
-      _ <- projectRepository.deactivateProjectById(projectId)
-      getProject <- projectRepository.getProjectById(projectId)
-    } yield getProject
+  def deactivateProjectById(projectId: ProjectId, userId: UserId): Future[Option[Project]] = {
+    val result = projectRepository.getProjectById(projectId)
+    result.flatMap {
+      case Some(project) if project.ownerId != userId => throw new ProjectAccessDeniedException
+      case Some(_) =>
+        for {
+          _ <- projectRepository.deactivateProjectById(projectId)
+          getProject <- projectRepository.getProjectById(projectId)
+        } yield getProject
+      case None => throw new ProjectNotFoundException
+    }
+  }
+
+  def updateProject(request: ProjectUpdateRequest, userId: UserId): Future[Int] =
+    projectRepository
+      .getProjectById(request.projectId)
+      .flatMap(
+        projectOpt =>
+          projectOpt.map(
+            project =>
+              if (project.ownerId == userId)
+                projectRepository.updateProject(
+                  project.copy(name = request.name, repository = request.repository)
+                )
+              else Future.failed(new ProjectAccessDeniedException)
+          ) match {
+            case Some(value) => value
+            case None        => Future.failed(new ProjectNotFoundException)
+          }
+      )
 
 }
+
+case class ProjectNotFoundException(message: String = "Project not found") extends RuntimeException(message)
+case class ProjectAccessDeniedException(message: String = "Access denied") extends RuntimeException(message)
```

Comment:
Not the best place for exceptions. Move it to the separate object, please.

Reply:
done